### ISSUE TITLE
libacvp cant  accsee to demo server

### ISSUE THREAD

**Issue by hexiangqwqwqw** [cisco/libacvp#126]
I am test libacvp. Please help me to check the log that I run the app of  libacvp to access to demo server . Do you have any opinion to slove this problem 

**root@hexiang-VirtualBox:/work/lib/libacvp# ./acvp_app 

Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: acvp/home
    ACV_CA_FILE:    certs/acvp.nist.gov.crt
    ACV_CERT_FILE:  kevin.micciche_hpe_server.cer
    ACV_KEY_FILE:   myserver.key

POST [
    {
        "acvVersion": "0.4"
    },
    {
        "password": "49464120"
    }
]
***ACVP [ERR][acvp_curl_http_post:265]--> Curl failed with code 58 (Problem with the local SSL certificate)

***ACVP [ERR][acvp_curl_http_post:281]--> HTTP response: 0

***ACVP [ERR][acvp_send_login:568]--> Unable to register with ACVP server. curl rv=0

***ACVP [STATUS][acvp_register:994]--> Login Response Failed (null)
***ACVP [STATUS][acvp_register:997]--> Login Send Failed
Failed to register with ACVP server (rv=3)**

**Comment by bfussell** [comment]
The only time I’ve seen that error is when I did not have the local cert/key files located in
the proper location.  Your log shows:


ACV_CA_FILE:    certs/acvp.nist.gov.crt

ACV_CERT_FILE:  kevin.micciche_hpe_server.cer
ACV_KEY_FILE:   myserver.key


Are you certain that your ACV_CERT_FILE and ACV_KEY_FILE are stored in your top level directory ?


Regards,

Barry

**Comment by hexiangqwqwqw** [comment]
I stroe the ACV_CERT_FILE and ACV_KEY_FILE in my top level directory. But the other error happenned , Please check the log:

root@hexiang-VirtualBox:/work/lib/libacvp# ./acvp_app 

Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: acvp
    ACV_CA_FILE:    certs/acvp.nist.gov.crt
    ACV_CERT_FILE:  kevin.micciche_hpe_server.cer
    ACV_KEY_FILE:   myserver.key


POST [
    {
        "acvVersion": "0.4"
    },
    {
        "password": "24345327"
    }
]
***ACVP [ERR][acvp_curl_http_post:281]--> HTTP response: 404

***ACVP [ERR][acvp_send_login:568]--> Unable to register with ACVP server. curl rv=404

***ACVP [STATUS][acvp_register:994]--> Login Response Failed <html><head><title>JBWEB000065: HTTP Status 404 - /acvpvalidation/acvp/login</title><style></style> </head><body><h1>JBWEB000065: HTTP Status 404 - /acvpvalidation/acvp/login</h1><HR size="1" noshade="noshade"><p><b>JBWEB000309: type</b> JBWEB000067: Status report</p><p><b>JBWEB000068: message</b> <u>/acvpvalidation/acvp/login</u></p><p><b>JBWEB000069: description</b> <u>JBWEB000124: The requested resource is not available.</u></p><HR size="1" noshade="noshade"></body></html>
***ACVP [STATUS][acvp_register:997]--> Login Send Failed
Failed to register with ACVP server (rv=3)

**Comment by hexiangqwqwqw** [comment]
I have solved the problem . But the KAT values always not ready

6 ***ACVP [STATUS][acvp_send_register:507]--> Successfully received registration response from ACVP server
527 ***ACVP [STATUS][acvp_register:1030]--> 200 OK [{"acvVersion":"0.4"},{"capabilityResponse":{"vectorSets":[{"vsId":2648},{"vsId":2649},{"vsId":2650},{"vsId":2
528 ***ACVP [STATUS][acvp_parse_register:1204]--> JWT: eyJhbGciOiJIUzI1NiJ9.eyJqdGkiOiI1NWIwNWVmYS03YTM4LTQzOTMtODY1ZC02ODJhNWZhN2NkZDgiLCJpYXQiOjE1MzQyMTMyOTksI
529 ***ACVP [STATUS][acvp_retrieve_vector_set:525]--> GET acvp/validation/acvp/vectors?vsId=2648
530 ***ACVP [STATUS][acvp_retrieve_vector_set:549]--> KAT vector set response received
531 ***ACVP [STATUS][acvp_process_vsid:1384]--> 200 OK [ {^M
532   "acvVersion" : "0.4"^M
533 }, {^M
534   "vsId" : 2648,^M
535   "retry" : 30^M
536 } ]
537  
538 ***ACVP [STATUS][acvp_retry_handler:1269]--> 200 OK KAT values not ready, server requests we wait and try again...

**Comment by bfussell** [comment]
The client should keep trying until the server sends the vectors.
What algorithms did you register for ?

**Comment by amrithathorath** [comment]
I'm having a similar issue:

> ***ACVP [ERR][acvp_curl_http_post:264]--> Curl failed with code 60 (Peer certificate cannot be authenticated with given CA certificates)
> ***ACVP [ERR][acvp_curl_http_post:280]--> HTTP response: 0
> ***ACVP [ERR][acvp_send_register:487]--> Unable to register with ACVP server. curl rv=0
> ***ACVP [ERR][acvp_send_register:488]--> (null)
> Failed to register with ACVP server (rv=3)

My values are as follows:

> export ACV_URI_PREFIX=acvp/
> export ACV_CA_FILE=certs/acvp.nist.gov.crt
> export ACV_KEY_FILE=privateKey.pem
> export ACV_CERT_FILE=certs/CSR_mycert.cer
> export ACV_PORT=443
> export ACV_SERVER=demo.acvts.nist.gov

Curl version is:

> $ curl -V
> curl 7.59.0 (x86_64-pc-linux-gnu) libcurl/7.59.0 OpenSSL/1.0.2j zlib/1.2.8
> Release-Date: 2018-03-14
> Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
> Features: AsynchDNS IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy
> 

Exported ssl path using variable LD_LIBRARY_PATH

**Comment by bfussell** [comment]
That CA file does not work for me, try the mozilla file in the certs directory

**Comment by amrithathorath** [comment]
@bfussell If I use the mozilla file you mentioned, should the demo server be demo.acvts.nist.gov or acvp.nist.gov?

**Comment by bfussell** [comment]
Here are my environment variables other than cert and key

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: acvp/
    ACV_CA_FILE:    certs/mozzila_trust_anchors.pem

**Comment by amrithathorath** [comment]
Checked paths to my certs. Seem to be correct!

**Comment by bfussell** [comment]
ACV_URI_PREFIX: acvp
needs to be
ACV_URI_PREFIX: acvp/

**Comment by amrithathorath** [comment]
A different error with that update:

```
amritha@thorath-vm:~/cavp_tool/libacvp-master$ ./app/acvp_app

Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: acvp/
    ACV_CA_FILE:    certs/mozzila_trust_anchors.pem
    ACV_CERT_FILE:  certs/CSR_athorath-corsec.cer
    ACV_KEY_FILE:   certs/privateKey.pem


POST [
    {
        "acvVersion": "0.4"
    },
    {
        "password": "56641647"
    }
]
***ACVP [ERR][acvp_curl_http_post:283]--> HTTP response: 401

***ACVP [ERR][acvp_send_login:616]--> Unable to register with ACVP server. curl rv=401

***ACVP [STATUS][acvp_register:1069]--> Login Response Failed {
  "acvVersion" : "0.4",
  "error" : "The supplied one time password could not be verified"
}
***ACVP [STATUS][acvp_register:1072]--> Login Send Failed
Failed to register with ACVP server (rv=3)

```

**Comment by bfussell** [comment]
Did you get an OTP from NIST ?
Are you certain that the device you are using for the client does not have too much time drift ?

**Comment by amrithathorath** [comment]
@bfussell I had received a zipped file with a CSR and a .txt file which I believe is TOTP seed. But I received this on 04/23, so it's been a while!

**Comment by bfussell** [comment]
I recommend checking with NIST, I’m pretty certain they have required new passwords since then.

**Comment by amrithathorath** [comment]
Will do thanks!!!

**Comment by amrithathorath** [comment]
@bfussell : Thanks for the tip regarding requesting new password from NIST. That seemed to solve the issue. I can successfully connect to the server now.  Now I see this message " 200 OK KAT values not ready, server requests we wait and try again..." which is similar to what is mentioned in one of the comments above. If I'm trying to run SHA for example, how do I register for it?

**Comment by bfussell** [comment]
You can run SHA only by using the command line options such as --no_aes, --no_tdes, etc. See the help information acvp_app --help

**Comment by amrithathorath** [comment]
@bfussell Is that same as registering for a particular alg?

**Comment by bfussell** [comment]
Yes, the help shows you how to enable and disable individual algorithms.
Have you modified app_main.c at all ? I'm not aware of any issues with SHA at this time.

**Comment by amrithathorath** [comment]
I don't think I modified app_main.c...

**Comment by bfussell** [comment]
Are you using branch V0.4 ? I just tested that and it works.

**Comment by amrithathorath** [comment]
Yes, I tried v0.4...
This is what I see.

> ***ACVP [STATUS][acvp_retry_handler:1349]--> 200 OK KAT values not ready, server requests we wait and try again...
> ***ACVP [STATUS][acvp_retrieve_vector_set:564]--> GET acvp/validation/acvp/vectors?vsId=10209
> ***ACVP [STATUS][acvp_retrieve_vector_set:588]--> KAT vector set response received
> ***ACVP [STATUS][acvp_process_vsid:1475]--> 200 OK [ {
>   "acvVersion" : "0.4"
> }, {
>   "vsId" : 10209,
>   "algorithm" : "SHA-1",
>   "testGroups" : [ {
>     "testType" : "AFT",
>     "tests" : [ {
>       "tcId" : 1,
>       "msg" : "",
>       "len" : 0
>     }, {
> ....
> ....
> ***ACVP [STATUS][acvp_dispatch_vector_set:1543]--> vsId: 10209
> ***ACVP [STATUS][acvp_dispatch_vector_set:1544]--> ACV Operation: SHA-1
> Error: Unsupported hash algorithm requested by ACVP server
> ***ACVP [ERR][acvp_hash_kat_handler:392]--> crypto module failed the operation

It says vector set received okay but in the end it says crypto module failed operation...

**Comment by fliphil** [comment]
@amrithathorath This will be fixed in #168. Also please open a new issue per problem that you have in the future so that we can properly reference them. Instead of 4 separate questions in 1 issue.

Thanks,
Philip

**Comment by elliemdaw** [comment]
Closing as #168 was merged.
